### ISSUE TITLE
GitLab authentication doesn't work with 2FA enabled in GitLab 8.11+

### ISSUE THREAD

**Issue by davejamesmiller** [composer/composer#5630]
With the following composer.json:

``` json
{
    ...
    "repositories": [
        {
            "type": "vcs",
            "url": "git@gitlab.com:myuser/myrepo.git"
        }
    ],
    "require": {
        "myuser/myrepo": "dev-master",
    },
    ...
}
```

Where `myuser/myrepo` is a private GitLab.com repo...

When I run this command:

```
composer install -vvv
```

I get this output:

```
Reading ./composer.json
Loading config file ./composer.json
Loading config file /home/www/sample/repo/auth.json
Reading /home/www/sample/repo/auth.json
Checked CA file /etc/ssl/certs/ca-certificates.crt: valid
Executing command (/home/www/sample/repo): git branch --no-color --no-abbrev -v
Reading /home/dave/.composer/composer.json
Loading config file /home/dave/.composer/composer.json
Reading /home/www/sample/repo/vendor/composer/installed.json
Reading /home/dave/.composer/packages/vendor/composer/installed.json
Running 1.2.0 (2016-07-19 01:28:52) with PHP 5.5.9-1ubuntu4.19 on Linux / 4.4.0-34-generic
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug
Reading ./composer.lock
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Reading ./composer.lock
Resolving dependencies through SAT
Dependency resolution completed in 0.006 seconds
Analyzed 136 packages to resolve dependencies
Analyzed 303 rules to resolve dependencies
  - Installing myuser/myrepo (dev-master 6804401)
Downloading https://gitlab.com/api/v3/projects/myuser%2Fmyrepo/repository/archive.zip?sha=6804401d74677c83c61c0f08c5cb5bcd6ad425a2
    Downloading: Connecting...
Could not fetch https://gitlab.com/api/v3/projects/myuser%2Fmyrepo/repository/archive.zip?sha=6804401d74677c83c61c0f08c5cb5bcd6ad425a2, enter your gitlab.com credentials to access private repos
A token will be created and stored in "/home/www/sample/repo/auth.json", your password will never be stored
To revoke access to this token you can visit gitlab.com/profile/applications
Username: myuser
Password:
Downloading https://gitlab.com/oauth/token
Bad credentials.
You can also manually create a personal token at https://gitlab.com/profile/applications
Add it using "composer config gitlab-oauth.gitlab.com <token>"
Username: ^C
```

I tried entering my account password, as I used to, but that seems to have been [disabled in GitLab 8.11](http://docs.gitlab.com/ee/api/oauth2.html) for security reasons:

> Starting in GitLab 8.11, the Resource Owner Password Credentials has been disabled for users with two-factor authentication turned on.

Then I tried entering a Personal Access Token from `https://gitlab.com/profile/personal_access_tokens`, and then my Private Token from `https://gitlab.com/profile/account`, but they didn't work either. (My guess is because these can only be used in a field named [`private_token`](http://docs.gitlab.com/ee/api/README.html#personal-access-tokens) not `password`.)

I also tried putting each of the tokens into `composer config gitlab-oauth.gitlab.com <token>` - again they didn't work, it just asked for my username and password again.

Finally I tried creating an OAuth application at `https://gitlab.com/profile/applications`, as linked in the message, but neither the application ID nor the secret key that it provides seem to be the right token to use in `composer config gitlab-oauth.gitlab.com <token>`.

For anyone else that runs into this problem, I found these temporary workarounds:
1. After 5 failed authentication attempts it falls back to installing the package from source via Git+SSH (so just hit Enter 10 times in a row).
2. Install with `composer install -n` to disable the authentication prompts and immediately fall back to installing from source.
3. If it's your own project, update `composer.json` to permanently disable API use - just change the type from `vcs` to `git`:
   
   ``` json
   {
       ...
       "repositories": [
           {
               "type": "git",
               "url": "git@gitlab.com:myuser/myrepo.git"
           }
       ],
       ...
   }
   ```
   
   Then run `composer update myuser/myrepo`.

When running `composer update`, rather than `composer install`, only the last option will work.

**Comment by curry684** [comment]
This is not a Composer issue, it's a longstanding issue on Gitlab's end that was finally fixed in 8.11: that you could perform many actions without 2FA despite having it enabled. The solution, like with many other applications [including Github](https://help.github.com/articles/creating-an-access-token-for-command-line-use/) lies with personal access tokens.

You should create an application access token for Composer at https://gitlab.com/profile/personal_access_tokens (or your own domain) and use that as your password instead.

**Comment by davejamesmiller** [comment]
@curry684 As I described above, I did try creating a personal access token and using that in place of my password, but that does not seem to work.

```
Username: myuser
Password: <entered personal access token>
Bad credentials.   
```

**Comment by curry684** [comment]
My bad, missed that. I am using similar setups with a private Gitlab instance, but using Satis in the middle to avoid exactly these issues, I find it more resilient.

**Comment by davejamesmiller** [comment]
Not a bad idea - I might look into that some time. :smiley:

**Comment by Chapoton** [comment]
Same issue here. Disabled 2FA and everything worked as expected.
Using GitLab Enterprise Edition 8.12.3-ee

**Comment by davejamesmiller** [comment]
@alcohol I'm pretty sure this should be labelled Bug rather than Support...

**Comment by alcohol** [comment]
@davejamesmiller I'm not yet convinced there is a bug to be honest. Might be our Gitlab integration is simply not up to date. But it basically depends on contributions from the community, since we do not really use it ourselves (we use Github as you can see).

**Comment by davejamesmiller** [comment]
> Might be our Gitlab integration is simply not up to date.

Not entirely sure what you mean there... If you mean it hasn't been updated to be compatible with the current GitLab API, then I'd agree - but is being incompatible with it not classed as a bug?

> But it basically depends on contributions from the community, since we do not really use it ourselves (we use Github as you can see).

That's fair enough.

**Comment by stof** [comment]
Use ``composer config gitlab-oauth.gitlab.com <token>`` with your personal access token.

The interactive creation of a token by asking you your username and password indeed does not accept personal tokens (personal tokens are not accepted to create other personal tokens) and composer does not support bitbucket's 2FA there.
But the composer output at this point explicitly gives the instructions to use a personal access token and configure composer to use it.

**Comment by davejamesmiller** [comment]
Hi @stof,

Thanks for your suggestion. However, I tried that already, as described above. Just in case anything changed since August, I tried it again just now on a new VM with a new repo - but it's still not working for me:

```
$ composer --version
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug
Composer version 1.2.3 2016-12-01 14:33:53

$ cat composer.json
{
    "repositories": [
        {
            "type": "vcs",
            "url": "git@gitlab.com:davejamesmiller/composer-gitlab-test-package.git"
        }
    ],
    "require": {
        "davejamesmiller/composer-gitlab-test-package": "dev-master"
    }
}

$ composer install
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug
Loading composer repositories with package information

Could not fetch https://gitlab.com/api/v3/projects/davejamesmiller%2Fcomposer%2Dgitlab%2Dtest%2Dpackage, enter your gitlab.com credentials to access private repos
A token will be created and stored in "/home/dave/composer-gitlab-test-project/auth.json", your password will never be stored
To revoke access to this token you can visit gitlab.com/profile/applications
Username: ^C

$ composer config gitlab-oauth.gitlab.com xxxxxxxxx_xxxxxxxxx_
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug

$ composer config gitlab-oauth.gitlab.com
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug
xxxxxxxxx_xxxxxxxxx_

$ composer install
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug
Loading composer repositories with package information

Could not fetch https://gitlab.com/api/v3/projects/davejamesmiller%2Fcomposer%2Dgitlab%2Dtest%2Dpackage, enter your gitlab.com credentials to access private repos
A token will be created and stored in "/home/dave/composer-gitlab-test-project/auth.json", your password will never be stored
To revoke access to this token you can visit gitlab.com/profile/applications
Username: ^C
```

(Replacing `xxxxxxxxx_xxxxxxxxx_` with the token I generated at https://gitlab.com/profile/personal_access_tokens)

FWIW, I verified that I can access https://gitlab.com/api/v3/projects/davejamesmiller%2Fcomposer%2Dgitlab%2Dtest%2Dpackage in a browser while logged in.

Any other ideas?
Thanks!

**Comment by davejamesmiller** [comment]
According to https://github.com/composer/composer/issues/5988#issuecomment-268536510 and https://github.com/composer/composer/issues/5661#issuecomment-268850814 this will be fixed in v1.3 so I'm closing this now.
